ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -67,10 +69,34 @@ public BlobId(String id, ClusterMap clusterMap)
    */
   public BlobId(DataInputStream stream, ClusterMap clusterMap)
       throws IOException {
+    this(stream, clusterMap, false);
+  }
+
+  /**
+   * Re-constructs existing blobId by deserializing from data input stream. This constructor includes an optional check
+   * that the stream has no more available bytes after reading.
+   *
+   * @param stream from which to deserialize the blobid
+   * @param clusterMap of the cluster that the blob id belongs to
+   * @param ensureFullyRead {@code true} if the stream should have no more available bytes after deserializing the blob
+   *                        ID.
+   * @throws IOException
+   */
+  private BlobId(DataInputStream stream, ClusterMap clusterMap, boolean ensureFullyRead)
+      throws IOException {
     this.version = stream.readShort();
     if (version == 1) {
-      this.partitionId = clusterMap.getPartitionIdFromStream(stream);
+      partitionId = clusterMap.getPartitionIdFromStream(stream);
+      if (partitionId == null) {
+        throw new IllegalArgumentException("Partition ID cannot be null");
+      }
       uuid = Utils.readIntString(stream);
+      if (uuid == null) {
+        throw new IllegalArgumentException("Could not parse uuid");
+      }
+      if (ensureFullyRead && stream.available() > 0) {
```

Comment:
I'm not sure this is the right thing to do. There are a couple problems with this according to me:

1) `stream.available()` is unreliable.
2) Blob IDs can be part of longer streams - for e.g. [here](https://github.com/linkedin/ambry/blob/master/ambry-protocol/src/main/java/com.github.ambry.protocol/MessageInfoListSerde.java#L74) and [here](https://github.com/linkedin/ambry/blob/master/ambry-commons/src/main/java/com.github.ambry.commons/BlobIdFactory.java#L38) (This is used whenever a record on disk is deserialized). I might be missing something here but are you a 100% sure this is the right thing to do? It doesn't seem so to me.
3) While I agree with the null checks, I'm not sure if we need to check if the stream has ended (at least inside blob id). Is it bad that we ignored a few characters? 
4) Ignoring (3), if you really want to do this check, I believe the context to check whether the stream has ended is with the caller of the constructor (the `Router` is where you want to fix this I'm guessing) and that is the component that needs to check that the stream has ended.

What do you think?